document-store/src/main/java/org/hypertrace/core/documentstore/Collection.java

```diff
@@ -16,6 +17,21 @@
    */
   boolean upsert(Key key, Document document) throws IOException;
 
+  /**
+   * Upsert (create a new doc or update if one already exists) the given document into the doc store
+   * if condition is evaluated to true. Provides optimistic lock support for concurrency update.
+   *
+   * @param key Unique key of the document in the collection.
+   * @param document Document to be upserted.
+   * @param condition Filter condition to be evaluated, on success update the document
+   * @param isUpsert Optional parameter to explicitly control insert or update, default is true.
+   *                 True indicates if the document doesn't exist, it will insert a new document
+   *                 False indicates if the document exists, update it otherwise don't do anything.
+   * @return True if success. False otherwise.
+   */
+  boolean upsert(Key key, Document document, Filter condition, @Nullable Boolean isUpsert)
```

Comment:
`final` args - to differentiates between inputs and outputs

Reply:
If the document exists, should it replace or merge? What if the merge is destructive?

Reply:
Primitive arguments (rather than a nullable Boolean) to clarify intention

Reply:
`isUpsert` is a redundant field - the api is called "upsert". If the user wants this distinction we should create separate APIs